thewalrus/_permanent.py

```diff
@@ -68,15 +61,96 @@ def perm(A, quad=True, fsum=False, method="bbfg"):
 
     isRyser = bool(method != "bbfg")
 
-    if A.dtype == np.complex:
-        if np.any(np.iscomplex(A)):
-            return perm_complex(A, quad=quad) if isRyser else perm_BBFG_complex(A)
-        return (
-            perm_real(np.float64(A.real), quad=quad, fsum=fsum)
-            if isRyser
-            else perm_BBFG_real(np.float64(A.real))
-        )
-    return perm_real(A, quad=quad, fsum=fsum) if isRyser else perm_BBFG_real(A)
+    return perm_ryser(A) if isRyser else perm_bbfg(A)
+
+
+@jit(nopython=True)
+def perm_ryser(M):  # pragma: no cover
+    """
+    Returns the permanent of a matrix using the Ryser formula in Gray ordering.
+
+    The code is an re-implementation from a Python 2 code found in
+    `Permanent code golf
+    <https://codegolf.stackexchange.com/questions/97060/calculate-the-permanent-as-quickly-as-possible>`_
+    using Numba.
+
+    Args:
+        M (array) : a square array.
+
+    Returns:
+        np.float64 or np.complex128: the permanent of matrix M.
+    """
+    # Raises an error if the matrix is not square
+    (a, b) = M.shape
+    if a != b:
+        raise Exception("Not a square matrix")
+    n = len(M)
```

Comment:
@josh146 : how do feel about raising exception from numba code? Note that these checks are also done by wrapper function.

Reply:
I have no great opinion 🙂 If the checks are already being done by the wrapper function, makes sense to remove it from here though I would think? Saves the validation being done twice.

Reply:
Note though it's best not to raise a bare `Exception`, but to always use a subclass e.g., `ValueError`